### PR TITLE
Implement guest side verification

### LINKED ISSUE #1306: Implement guest side proof verification
`ZkVm` trait is implemented for adapters in our repo, left with todo!()'s
 
 for risc0 we need to also modify `ZkVmHost` because there is two types of inputs for the circuit; inputs and assumptions. and their recursion api only supports verification through assumptions.
 
 TODOs:
- [x] Clear up guest module under adapters/risc0, Risc0Guest was written a little bit complicated.
- [x] Modify `ZkVmHost` trait and add `add_assumption` method.
- [x] Implement `add_assumption` for Risc0Host and Risc0BonsaiHost
- [x] Implement `ZkVm::verify` and `ZkVm::verify_extract_output` functions for Risc0Guest
- [x] Figure out how to make verification work in execute mode. Risc0 checks some env vars that make mock verification.

### LINKED ISSUE #1392: Merge fake proof and full proof
We can merge Proof enum into one type since now both Fake and Full proof are verifiable

### PR DESCRIPTION
**Pull request by ercecan**
# Description
This pr supersedes #1375  and #1385 
Implements guest side verification for light client prover

## Linked Issues
- Fixes #1306 
- Fixes #1392

### CODE REVIEW

File: asdasd.txt
Comment (eyusufatik) [inline asdasd.txt:1]: looool

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: asdasd.txt
Comment (ercecan) [reply asdasd.txt:1]: :D:D:D

File: crates/light-client-prover/src/circuit.rs [hunk 1]
```diff
@@ -42,9 +45,17 @@ pub fn run_circuit<DaV: DaVerifier>(
         }
     }
 
+    let batch_proof_journals = input.batch_proof_journals.clone();
```
Comment (eyusufatik) [inline crates/light-client-prover/src/circuit.rs]: you don't need clone() here

File: crates/light-client-prover/src/da_block_handler.rs [hunk 2]
```diff
@@ -149,10 +149,27 @@ where
 
         // Do any kind of ordering etc. on batch proofs here
         // If you do so, don't forget to do the same inside zk
+        let batch_proof_method_id = self
+            ._batch_proof_code_commitments_by_spec
```
Comment (eyusufatik) [inline crates/light-client-prover/src/da_block_handler.rs]: remove _ in var name

File: crates/risc0-bonsai/src/host.rs [hunk 3]
```diff
@@ -96,6 +98,7 @@ impl<'a> Risc0BonsaiHost<'a> {
         // Compute the image_id, then upload the ELF with the image_id as its key.
         // handle error
         let image_id = compute_image_id(elf).unwrap();
+        println!("computed image id: {:?}", image_id);
```
Comment (eyusufatik) [inline crates/risc0-bonsai/src/host.rs]: remove

File: crates/risc0-bonsai/src/host.rs [hunk 4]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
         Risc0Guest::with_hints(std::mem::take(&mut self.env))
     }
 
+    fn add_assumption(&mut self, receipt_buf: Vec<u8>) {
+        let receipt: Receipt = bincode::deserialize(&receipt_buf).expect("Receipt should be valid");
+        let ass: AssumptionReceipt = receipt.into();
```
Comment (eyusufatik) [inline crates/risc0-bonsai/src/host.rs]: rename

Comment (eyusufatik) [reply crates/risc0-bonsai/src/host.rs]: I know it's short for assumption but ti doesn't look good.

File: crates/risc0-bonsai/src/host.rs [hunk 5]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
+        let a = ass.clone().claim_digest().unwrap();
+        println!("Digest of assumption: {:?}", a);
```
Comment (eyusufatik) [inline crates/risc0-bonsai/src/host.rs]: remove print

File: crates/risc0-bonsai/src/host.rs [hunk 6]
```diff
@@ -449,10 +477,20 @@ impl<'host> Zkvm for Risc0BonsaiHost<'host> {
         code_commitment: &Self::CodeCommitment,
     ) -> Result<Vec<u8>, Self::Error> {
         let receipt: Receipt = bincode::deserialize(serialized_proof)?;
+        let cl = receipt.claim().unwrap();
```
Comment (eyusufatik) [inline crates/risc0-bonsai/src/host.rs]: remove

File: crates/risc0-bonsai/src/host.rs [hunk 7]
```diff
@@ -449,10 +477,20 @@ impl<'host> Zkvm for Risc0BonsaiHost<'host> {
...
+        let dg = cl.digest();
+        tracing::error!("receipt digest: {:?}", dg);
+        tracing::error!("receipt digest bytes: {:?}", dg.as_bytes());
+        tracing::error!("receipt digest words: {:?}", dg.as_words());
 
         #[allow(clippy::clone_on_copy)]
         receipt.verify(code_commitment.clone())?;
 
+        println!(
```
Comment (eyusufatik) [inline crates/risc0-bonsai/src/host.rs]: remove

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/light-client-prover/src/da_block_handler.rs [hunk 8]
```diff
@@ -149,10 +149,27 @@ where
...
+            .get(&SpecId::Genesis)
+            .expect("Batch proof code commitment not found");
+
+        let mut assumptions = vec![];
+        let mut journals = vec![];
+        for batch_proof in batch_proofs {
+            if let DaDataLightClient::Complete(proof) = batch_proof {
+                // Handle err
+                let output = Vm::verify(proof.as_slice(), batch_proof_method_id).unwrap();
```
Comment (yaziciahmet) [inline crates/light-client-prover/src/da_block_handler.rs]: I believe we don't necessarily need to verify here. We can just use extract_output or some similar method, no?

File: crates/light-client-prover/src/da_block_handler.rs [hunk 9]
```diff
@@ -197,30 +214,43 @@ where
...
         da_data: Vec<<<Da as DaService>::Spec as DaSpec>::BlobTransaction>,
         l1_block: &Da::FilteredBlock,
+        batch_prover_code_commitment: &<Vm as Zkvm>::CodeCommitment,
+        journals: Vec<Vec<u8>>,
     ) -> LightClientCircuitInput<Da::Spec> {
         let (inclusion_proof, completeness_proof) = self
             .da_service
             .get_extraction_proof_light_client(l1_block)
             .await;
+        println!(
```
Comment (yaziciahmet) [inline crates/light-client-prover/src/da_block_handler.rs]: remove

File: crates/prover-services/src/parallel/prover.rs [hunk 10]
```diff
@@ -93,13 +96,26 @@ where
...
+    ) {
+        let header_hash = da_slot_hash;
+        let mut prover_state = self.prover_state.lock();
+        let prover_status = prover_state.prover_status.get_mut(&header_hash);
+        if let Some(ProverStatus::WitnessSubmitted(data)) = prover_status {
+            data.1 = assumptions
+        }
+    }
+
+    pub(crate) fn submit_input(
```
Comment (yaziciahmet) [inline crates/prover-services/src/parallel/prover.rs]: I feel like these 2 are dangerous, as they can overwrite each other. I think behavior for these 2 methods should go like:
1. lock
2. if entry doesnt exist, create a default empty entry.
3. only update the assumption or input based on whatever is the method.

Maybe we don't use it in that manner at the current state in the code, but they shouldn't be able to overwrite each other imo. Specifically, submit_input -> submit_assumptions.

File: crates/risc0-bonsai/src/host.rs [hunk 11]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
+                }
+
+                let env = env.write_slice(&self.env).build().unwrap();
+
+                let prover = LocalProver::new("citrea-test-prover");
+
+                tracing::info!("Starting risc0 proving");
+                let ProveInfo { receipt, stats } = prover.prove(env, self.elf)?;
+
+                println!(
```
Comment (yaziciahmet) [inline crates/risc0-bonsai/src/host.rs]: remove

File: crates/risc0-bonsai/src/host.rs [hunk 12]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
+                );
+
+                // Because the dev mode is set, the proof will generate a fake receipt which does not include the proof
+                // It only includes the journal
+                assert!(matches!(receipt.inner, InnerReceipt::Fake(_)));
 
-                let session = executor.run()?;
-                let data =
-                    bincode::serialize(&session.journal.expect("Journal shouldn't be empty"))?;
+                tracing::warn!("Execution Stats: {:?}", stats);
```
Comment (yaziciahmet) [inline crates/risc0-bonsai/src/host.rs]: info!

File: crates/risc0-bonsai/src/host.rs [hunk 13]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
-                    .write_slice(&self.env)
-                    .build()
-                    .unwrap();
-                let mut executor = ExecutorImpl::from_elf(env, self.elf)?;
+                // Set dev mode to true as this is used for tests
+                std::env::set_var("RISC0_DEV_MODE", "1");
+                let mut env = add_benchmarking_callbacks(ExecutorEnvBuilder::default());
+                for assumption in self.assumptions.iter() {
+                    env.add_assumption(assumption.clone());
+                    tracing::info!("added assumption to the env");
```
Comment (yaziciahmet) [inline crates/risc0-bonsai/src/host.rs]: maybe debug, or trace, with more info about the assumption, like size of it?

File: crates/risc0-bonsai/src/host.rs [hunk 14]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
+
+                let env = env.write_slice(&self.env).build().unwrap();
+
+                let prover = LocalProver::new("citrea-test-prover");
+
+                tracing::info!("Starting risc0 proving");
+                let ProveInfo { receipt, stats } = prover.prove(env, self.elf)?;
+
+                println!(
+                    "\n\n\nreceipt journal after proving: {:?} \n\n\n",
```
Comment (rakanalh) [inline crates/risc0-bonsai/src/host.rs]: Is this a left over?
```suggestion
                    "receipt journal after proving: {:?}",
```

Or should the whole print be removed?

File: crates/risc0-bonsai/src/host.rs [hunk 15]
```diff
@@ -293,30 +295,62 @@ impl<'a> ZkvmHost for Risc0BonsaiHost<'a> {
...
+        let ass: AssumptionReceipt = receipt.into();
+        let a = ass.clone().claim_digest().unwrap();
+        println!("Digest of assumption: {:?}", a);
+
+        self.assumptions.push(ass);
+    }
+
     /// Only with_proof = true is supported.
     /// Proofs are created on the Bonsai API.
     fn run(&mut self, with_proof: bool) -> Result<Proof, anyhow::Error> {
```
Comment (yaziciahmet) [inline crates/risc0-bonsai/src/host.rs:304]: clean assumptions after a run as well.

**Review by yaziciahmet — DISMISSED** [review]

File: crates/prover-services/src/parallel/prover.rs (hunk 10 above)
Comment (ercecan) [reply crates/prover-services/src/parallel/prover.rs]: This was also on my mind, thanks for the heads up

File: crates/risc0-bonsai/src/host.rs (hunk 14 above)
Comment (ercecan) [reply crates/risc0-bonsai/src/host.rs]: Removed

File: crates/light-client-prover/src/da_block_handler.rs (hunk 8 above)
Comment (ercecan) [reply crates/light-client-prover/src/da_block_handler.rs]: Should we not verify the proofs at some point tho? Check out the new version

Comment (eyusufatik) [reply crates/light-client-prover/src/da_block_handler.rs]: we should verify imo 

so that we can seperate inputs accordingly

Comment (yaziciahmet) [reply crates/light-client-prover/src/da_block_handler.rs]: but we verify already in circuit. we just need the output.

one benefit would be to not start proving if we see beforehand that its broken. I am not sure what to do with that neither.

Comment (yaziciahmet) [reply crates/light-client-prover/src/da_block_handler.rs]: assertion here is fine imo, but expect instead of unwrap could be slightly better.

Comment (eyusufatik) [reply crates/light-client-prover/src/da_block_handler.rs]: > one benefit would be to not start proving if we see beforehand that its broken. I am not sure what to do with that neither.

we should verify for all sorts of reasons, trust me

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**